x-pack/plugin/search-engines/src/main/java/org/elasticsearch/searchengines/analytics/SearchEngineAnalyticsBuilder.java

```diff
@@ -0,0 +1,235 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.searchengines.analytics;
+
+import org.elasticsearch.ElasticsearchException;
+import org.elasticsearch.ResourceAlreadyExistsException;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.admin.indices.template.get.GetComponentTemplateAction;
+import org.elasticsearch.action.admin.indices.template.put.PutComponentTemplateAction;
+import org.elasticsearch.action.admin.indices.template.put.PutComposableIndexTemplateAction;
+import org.elasticsearch.action.datastreams.CreateDataStreamAction;
+import org.elasticsearch.action.datastreams.GetDataStreamAction;
+import org.elasticsearch.action.support.master.AcknowledgedResponse;
+import org.elasticsearch.client.internal.Client;
+import org.elasticsearch.client.internal.Requests;
+import org.elasticsearch.cluster.metadata.ComponentTemplate;
+import org.elasticsearch.cluster.metadata.ComposableIndexTemplate;
+import org.elasticsearch.cluster.metadata.Template;
+import org.elasticsearch.common.compress.CompressedXContent;
+import org.elasticsearch.logging.LogManager;
+import org.elasticsearch.logging.Logger;
+import org.elasticsearch.xcontent.XContentBuilder;
+import org.elasticsearch.xcontent.XContentFactory;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * The SearchRequestAnalyticsBuilder tries to ensure that an analytics data stream exists for the specified
+ * {@link org.elasticsearch.cluster.metadata.SearchEngine} if an analytics collection is specified.
+ * If for some reason a data stream cannot be created, we log a message but don't throw an error, and allow
+ * engine requests to continue.
+ */
+public class SearchEngineAnalyticsBuilder {
+
+    private static final Logger logger = LogManager.getLogger(SearchEngineAnalyticsBuilder.class);
+    private static final int MAX_RETRIES = 3;
+    private static final String ANALYTICS_TEMPLATE_POSTFIX = "_template";
+    private static final String COMPONENT_MAPPING_TEMPLATE_POSTFIX = "_mappings";
+
+    public static void ensureDataStreamExists(String dataStreamName, Client client) {
```

Comment:
We should avoid creating static method but an instance instead.
This instance should be instantiated in the createComponents method of the plugin.

Also, client should not be a method parameter but a property of the instance.

```java
public class SearchEngineAnalyticsBuilder {
  private final Client client;

  public SearchEngineAnalyticsBuilder(Cllient client) {
    this.client = client
  }

  public static void ensureDataStreamExists(String dataStreamName) {
    ...
  }
}
```

createComponents code should be:

```java
        var searchEngineAnalyticsBuilder = new SearchEngineAnalyticsBuilder(client);
        var searchEngineMetadataService = new SearchEngineMetadataService(clusterService, searchEngineAnalyticsBuilder);
        return List.of(searchEngineMetadataService);
```

PS: I try to avoid static methods as much as possible since it makes it harder to write test using mocks.

Reply:
> PS: I try to avoid static methods as much as possible since it makes it harder to write test using mocks.

I like to use them when possible if they do not depend on object state. But of course, IT.depends